### PR TITLE
Add lease

### PR DESCRIPTION
**Pull request by ldpliu**
https://github.com/open-cluster-management/backlog/issues/6893
https://github.com/open-cluster-management/backlog/issues/7130
https://github.com/open-cluster-management/backlog/issues/7067

### CODE REVIEW

File: pkg/klusterlet/lease/lease_controller.go [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+		}
+	}
+
+	return
+}
+
+//checkPodIsRunning check if the pod is ready
+func (r *LeaseReconciler) checkPodIsRunning() (bool, error) {
+	if r.PodName == "" || r.PodNamespace == "" {
+		return true, nil
```
Comment (zhiweiyin318) [inline pkg/klusterlet/lease/lease_controller.go:122]: should return false here, right ?

Comment (ldpliu) [reply pkg/klusterlet/lease/lease_controller.go]: yes, you are right. changed it.

File: pkg/klusterlet/lease/lease_controller.go [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+			labelSelector := labels.FormatLabels(curPod.Labels)
+			err = r.KubeClient.CoreV1().Pods(r.PodNamespace).DeleteCollection(context.TODO(), metav1.DeleteOptions{}, metav1.ListOptions{LabelSelector: labelSelector})
+			if err != nil {
+				klog.Errorf("failed to restart pod. error:%v", err)
+				return
+			}
+		}
+	}
+
+	hubClient, err := r.BuildKubeClientWithSecretFunc(r.cachedSecret)
```
Comment (qiujian16) [inline pkg/klusterlet/lease/lease_controller.go:74]: I did not see where the cachesecret  is updated, Do I miss that code?

File: pkg/klusterlet/lease/lease_controller.go [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+	}
+	lease.Spec.RenewTime = &metav1.MicroTime{Time: time.Now()}
+	if _, err = hubClient.CoordinationV1().Leases(r.LeaseNamespace).Update(context.TODO(), lease, metav1.UpdateOptions{}); err != nil {
+		klog.Errorf("unable to update cluster lease %q/%q on hub cluster. error:%v", r.LeaseNamespace, r.LeaseName, err)
+		return
+	}
+	return
+}
+
+func (r *LeaseReconciler) checkHubSecret(instance *corev1.Secret) bool {
```
Comment (qiujian16) [inline pkg/klusterlet/lease/lease_controller.go:112]: I am not  sure why  you need this, you have checked this in reconcile.

File: pkg/klusterlet/lease/lease_controller.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+		r.hubClient, err = utils.BuildKubeClient(r.HubKubeConfigPath)
+		if err != nil {
+			klog.Errorf("failed to build hub client. error:%v", err)
+			return
+		}
+
+	}
+
+	lease, err := r.hubClient.CoordinationV1().Leases(r.LeaseNamespace).Get(context.TODO(), r.LeaseName, metav1.GetOptions{})
+	if err != nil {
```
Comment (qiujian16) [inline pkg/klusterlet/lease/lease_controller.go:74]: better to  use switch  here

switch {
errors.IsNotFound(err):
  create lease
  return
err != nil
  return
}

File: pkg/klusterlet/lease/lease_controller_test.go [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+	},
+}
+
+func TestLeaseReconciler_Reconcile(t *testing.T) {
+	s := scheme.Scheme
+	s.AddKnownTypes(corev1.SchemeGroupVersion, &corev1.Namespace{})
+	kubeClient = kubefake.NewSimpleClientset(ns)
+	kubeconfigData := []byte("aa")
+
+	tempdir, err := ioutil.TempDir("", "kube")
```
Comment (qiujian16) [inline pkg/klusterlet/lease/lease_controller_test.go:41]: i  do not think you would need  this part, the namespace can been set, right?

File: pkg/klusterlet/lease/lease_controller_test.go [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+	leaseReconciler1 := &LeaseReconciler{
+		KubeClient:           kubeClient,
+		LeaseName:            leaseName,
+		LeaseNamespace:       leaseNamespace,
+		LeaseDurationSeconds: 1,
+		cachedKubeConfig:     kubeconfigData,
+		HubKubeConfigPath:    "",
+		hubClient:            kubeClient,
+	}
+	leaseReconciler1.Reconcile(context.TODO())
```
Comment (qiujian16) [inline pkg/klusterlet/lease/lease_controller_test.go:87]: What is the purpose of these tests?

File: pkg/utils/utils.go [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/klog"
+)
+
+func GetComponentNamespace() (string, error) {
+	nsBytes, err := ioutil.ReadFile("/var/run/secrets/kubernetes.io/serviceaccount/namespace")
+	if err != nil {
+		return "", err
```
Comment (qiujian16) [inline pkg/utils/utils.go:14]: return open-cluster-management-agent by default

File: pkg/klusterlet/lease/lease_controller.go [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+			},
+		}
+		if _, err := r.hubClient.CoordinationV1().Leases(r.LeaseNamespace).Create(context.TODO(), lease, metav1.CreateOptions{}); err != nil {
+			klog.Errorf("unable to create addon lease %q/%q on hub cluster. error:%v", r.LeaseNamespace, r.LeaseName, err)
+		}
+		return
+	case err != nil:
+		klog.Errorf("unable to get addon lease %q/%q on hub cluster. error:%v", r.LeaseNamespace, r.LeaseName, err)
+		return
+	case err == nil:
```
Comment (qiujian16) [inline pkg/klusterlet/lease/lease_controller.go:93]: you do not need this case

File: pkg/klusterlet/lease/lease_controller.go [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+			labelSelector := labels.FormatLabels(agentLabel)
+			err = r.KubeClient.CoreV1().Pods(r.componentNamespace).DeleteCollection(context.TODO(), metav1.DeleteOptions{}, metav1.ListOptions{LabelSelector: labelSelector})
+			if err != nil {
+				klog.Errorf("failed to restart pod. error:%v", err)
+			}
+			return
+		}
+
+		//update cached kubeconfig and hub client
+		r.cachedKubeConfig = curKubeConfig
```
Comment (qiujian16) [inline pkg/klusterlet/lease/lease_controller.go:62]: the below part can still have error, so move line 62 after line 64

File: pkg/klusterlet/lease/lease_controller.go [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+			},
+		}
+		if _, err := r.hubClient.CoordinationV1().Leases(r.LeaseNamespace).Create(context.TODO(), lease, metav1.CreateOptions{}); err != nil {
+			klog.Errorf("unable to create addon lease %q/%q on hub cluster. error:%v", r.LeaseNamespace, r.LeaseName, err)
+		}
+		return
+	case err != nil:
+		klog.Errorf("unable to get addon lease %q/%q on hub cluster. error:%v", r.LeaseNamespace, r.LeaseName, err)
+		return
+	default:
```
Comment (qiujian16) [inline pkg/klusterlet/lease/lease_controller.go:93]: commonly you do not need  this default either...you can move the following code out  of switch block

**Pull request merged**